### ISSUE TITLE
guildai 0.7.0rc10 seems to cause multithreading issues

### ISSUE THREAD

**Issue by thomsentner** [guildai/guildai#197]
I'm running into problems after updating to 0.7.0rc10. I'm not having these same issues with 0.7.0rc9. I've installed both in a clean environment for comparison purposes.

My code runs until it starts executing sklearn's `RandomizedSearchCV()`. If I pass `n_jobs=6` to RandomizedSearchCV, only 0.7.0rc9 is able to run the code. When using `n_jobs=1` or `n_jobs=None`, both versions are able to run the code.

The error message I am getting:
```
joblib.externals.loky.process_executor.TerminatedWorkerError: A worker process managed by the executor was unexpectedly terminated. This could be caused by a segmentation fault while calling the function or by an excessive memory usage causing the Operating System to kill the worker.

The exit codes of the workers are {SIGSEGV(-11)}

```
Interestingely, with 0.7.0rc10 I'm getting the following warning right before my code starts:
```
INFO: [numexpr.utils] Note: NumExpr detected 24 cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 8.
INFO: [numexpr.utils] NumExpr defaulting to 8 threads.
```

Let me know if I can provide any additional details.

**Comment by gar1t** [comment]
Thanks a lot for the report! Could I trouble you for a small example that I can run that reproduces the behavior on your system?

Also, what is the Guild command that you run that triggers this? If the command references a Pipeline op in a Guild file (i.e. an operation that uses `steps`) I'd need to know what that looks like.

**Comment by thomsentner** [comment]
Sure, the following code consistently fails for me on 0.7.0rc10, though runs on 0.7.0rc9. I've written the code to be somewhat similar to what I actually use.

```
import xgboost as xgb
from sklearn.model_selection import RandomizedSearchCV, GroupKFold
import numpy as np

cv = RandomizedSearchCV(
    estimator=xgb.XGBClassifier(),
    param_distributions={"gamma": [0.5, 1]},
    cv=GroupKFold(n_splits=4),
    n_jobs=6,
    n_iter=1,
)

cv.fit(
    X=np.random.random_sample((1000, 100)),
    y=np.random.choice(3, 1000),
    groups=np.random.choice(25, 1000)
)
```

I'm using xgboost 1.0.2, numpy 1.18.5 and sklearn 0.22.2.post1.

EDIT:
Forgot to add, I'm just using `guild run script.py` to run my code.

**Comment by gar1t** [comment]
I created an issue resolution doc to help work on this:

https://github.com/guildai/issue-resolution/tree/master/197-guildai-0.7.0rc10-seems-to-cause-multithreading-issues

When I follow the steps-to-recreate in that doc I'm not able to see any issues. It appears to run the same as when I run with Python directly.

When you're able, could you clone that repo and see if you can reproduce the issue on your system?

**Comment by thomsentner** [comment]
Thanks. I've done as you instructed, and I am able to successfully run the code when using guild's virtual environment. 

I normally use conda myself to install most of my packages. I've tried to come up with as simple scenarios as possible where still rc10 fails but rc9 doesn't:

I first initialize an environment using:
```
conda create -n guildcondaenv python3.7
conda activate guildcondaenv
```
Followed by either of the following:
```
conda install -c conda-forge scikit-learn=0.22.2.post1 numpy=1.18.5
pip install guildai==0.7.0.rcxx xgboost==1.0.2
```
Above works for rc9 and rc10.
```
conda install -c conda-forge scikit-learn=0.22.2.post1 numpy=1.18.5 xgboost=1.0.2
pip install guildai==0.7.0.rcxx
```
Above only works for rc9.

<details>
  <summary>My lines exactly</summary>

```
conda create -n swb_xpg9 python=3.7 -y
conda create -n swb_xcg9 python=3.7 -y
conda create -n swb_xpg10 python=3.7 -y
conda create -n swb_xcg10 python=3.7 -y
conda activate swb_xpg9
conda install -y -c conda-forge scikit-learn=0.22.2.post1 numpy=1.18.5
pip install guildai==0.7.0.rc9 xgboost==1.0.2
guild run test.py -y
conda activate swb_xcg9
conda install -y -c conda-forge scikit-learn=0.22.2.post1 numpy=1.18.5 xgboost=1.0.2
pip install guildai==0.7.0.rc9
guild run test.py -y
conda activate swb_xpg10
conda install -y -c conda-forge scikit-learn=0.22.2.post1 numpy=1.18.5
pip install guildai==0.7.0.rc10 xgboost==1.0.2
guild run test.py -y
conda activate swb_xcg10
conda install -y -c conda-forge scikit-learn=0.22.2.post1 numpy=1.18.5 xgboost=1.0.2
pip install guildai==0.7.0.rc10
guild run test.py -y
# only last run fails
```
</details>


TBH this is a bit weirder than I expected. It seems to be related to xgboost. I am wondering if it could be the result of some dependencies.

EDIT:
To be clear, when executing `python test.py` in the environment where guild fails, the script finishes.

**Comment by gar1t** [comment]
Haha as you soon as you said 'thread' I knew it was going to be weird :) I'll try to reproduce using your details steps - thanks a lot for the careful examination.

**Comment by thomsentner** [comment]
Seems to be fixed in 0.7.0.rc11

**Comment by gar1t** [comment]
Closing for 0.7.0.
